src/client/src/database.rs

```diff
@@ -283,55 +288,110 @@ impl Database {
-            .and_then(|response| response.into_inner().try_collect())
-            .await
-            .map_err(|e| {
-                let tonic_code = e.code();
-                let e: error::Error = e.into();
-                let code = e.status_code();
-                let msg = e.to_string();
-                ServerSnafu { code, msg }
-                    .fail::<()>()
-                    .map_err(BoxedError::new)
-                    .context(error::FlightGetSnafu {
+        let response = client.mut_inner().do_get(request).await.map_err(|e| {
+            let tonic_code = e.code();
+            let e: error::Error = e.into();
+            let code = e.status_code();
+            let msg = e.to_string();
+            ServerSnafu { code, msg }
+                .fail::<()>()
+                .map_err(BoxedError::new)
+                .context(error::FlightGetSnafu {
+                    tonic_code,
+                    addr: client.addr(),
+                })
+                .map_err(|error| {
+                    logging::error!(
+                        "Failed to do Flight get, addr: {}, code: {}, source: {}",
+                        client.addr(),
                         tonic_code,
-                        addr: client.addr(),
-                    })
-                    .map_err(|error| {
-                        logging::error!(
-                            "Failed to do Flight get, addr: {}, code: {}, source: {}",
-                            client.addr(),
-                            tonic_code,
-                            error
-                        );
                         error
-                    })
-                    .unwrap_err()
-            })?;
-
-        let decoder = &mut FlightDecoder::default();
-        let flight_messages = flight_data
-            .into_iter()
-            .map(|x| decoder.try_decode(x).context(ConvertFlightDataSnafu))
-            .collect::<Result<Vec<_>>>()?;
-
-        let output = if let Some(FlightMessage::AffectedRows(rows)) = flight_messages.get(0) {
-            ensure!(
-                flight_messages.len() == 1,
-                IllegalFlightMessagesSnafu {
-                    reason: "Expect 'AffectedRows' Flight messages to be one and only!"
-                }
-            );
-            Output::AffectedRows(*rows)
-        } else {
-            let recordbatches = flight_messages_to_recordbatches(flight_messages)
-                .context(ConvertFlightDataSnafu)?;
-            Output::RecordBatches(recordbatches)
+                    );
+                    error
+                })
+                .unwrap_err()
+        })?;
+
+        let flight_data_stream = response.into_inner();
+        let mut decoder = FlightDecoder::default();
+
+        let mut flight_message_stream = flight_data_stream.map_ok(move |flight_data| {
+            decoder
+                .try_decode(flight_data)
+                .context(ConvertFlightDataSnafu)
+        });
```

Comment:
You can treat this stream using:

```suggestion
        let mut flight_message_stream = flight_data_stream.map(move |flight_data| {
            flight_data
                .map_err(|s| { error::Error::from(s) })
                .and_then(|data| decoder.try_decode(data).context(ConvertFlightDataSnafu))
        });
```

So that the stream yields a `Result<...>` instead of `Result<Result<...>>` and you don't have to match a nested result.

Reply:
`.map_err(|s| { error::Error::from(s) })` -> `.map_err(error::Error::from)`